kythe/java/com/google/devtools/kythe/analyzers/java/AutoValuePlugin.java

```diff
@@ -0,0 +1,296 @@
+    while (autoValueClass == null && classType.supertype_field instanceof ClassType) {
+      classType = (ClassType) classType.supertype_field;
+      JCTree superTypeTree = javacTrees.getTree(classType.asElement());
+      if (superTypeTree instanceof JCClassDecl) {
+        JCClassDecl cls = (JCClassDecl) superTypeTree;
+        if (getAnnotation(AUTO_VALUE_ANNOTATION, cls.getModifiers().getAnnotations()).isPresent()) {
+          autoValueClass = cls;
+        }
+      }
+    }
+    if (autoValueClass == null) {
+      return false;
+    }
+
+    // @AutoValue class generates the implementation class.
+    Optional<KytheNode> autoNode = kytheGraph.getNode(autoValueClass.sym);
+    Optional<KytheNode> genNode = kytheGraph.getNode(classDef.sym);
+    if (autoNode.isPresent() && genNode.isPresent()) {
+      entrySets.emitEdge(autoNode.get().getVName(), EdgeKind.GENERATES, genNode.get().getVName());
+    }
+
+    Map<String, JCMethodDecl> propMethods = new HashMap<>(); // property_name -> getter method
+    Map<String, VName> getterNodes = new HashMap<>(); // property_name -> Kythe getter VName
+    Map<String, VName> properties = new HashMap<>(); // property_name -> Kythe property node VName
+
+    // Find potential propMethods in @AutoValue class.
+    for (JCTree def : autoValueClass.getMembers()) {
+      if (def instanceof JCMethodDecl) {
+        JCMethodDecl method = (JCMethodDecl) def;
+        if (method.getModifiers().getFlags().contains(Modifier.ABSTRACT)
+            && method.getParameters().isEmpty()
+            && !isVoid(method.getReturnType())) {
+          Optional<KytheNode> kytheNode = kytheGraph.getNode(method.sym);
+          if (kytheNode.isPresent()) {
+            String name = method.sym.getSimpleName().toString();
+            // TODO(schroederc): possibly reuse getter method as "property" node or change node kind
+            EntrySet propNode =
+                entrySets.newNode("property").addSignatureSalt(kytheNode.get().getVName()).build();
+            propNode.emit(entrySets.getEmitter());
+            entrySets.emitEdge(
+                kytheNode.get().getVName(), EdgeKind.PROPERTY_READS, propNode.getVName());
+
+            properties.put(name, propNode.getVName());
+            propMethods.put(name, method);
+            getterNodes.put(name, kytheNode.get().getVName());
+          }
+        }
+      }
+    }
+    if (propMethods.isEmpty()) {
+      return false;
+    }
+
+    // Either all getters must be prefixed or all names are taken as written.
+    boolean allPrefixed =
+        properties.keySet().stream().allMatch(AutoValuePlugin::hasGetPropertyPrefix);
+    if (allPrefixed) {
+      Map<String, VName> additional = new HashMap<>();
+      for (Map.Entry<String, VName> prop : properties.entrySet()) {
+        additional.put(propertyName(prop.getKey()), prop.getValue());
+      }
+      properties.putAll(additional);
+    }
+
+    // Find potential @AutoValue.Builder and its defined setters.
```

Comment:
This method is huge. Maybe factor out the block below to a getBuilderClass() method? (you'll end up with one extra call to kytheGraph.getNode but so what).